bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerHandle.java

```diff
@@ -1898,302 +1788,81 @@ void handleBookieFailure(final Map<Integer, BookieSocketAddress> failedBookies)
-                    LOG.debug("[EnsembleChange-L{}-{}] : metadata is conflicted, local metadata = \n {} \n,"
-                        + " zk metadata = \n {} \n", ledgerId, ensembleChangeIdx, metadata, newMeta);
-                }
-                return false;
-            }
+    void ensembleChangeLoop(List<BookieSocketAddress> origEnsemble, Map<Integer, BookieSocketAddress> failedBookies) {
+        // when the ensemble changes are too frequent, close handle
+        if (numEnsembleChanges.incrementAndGet() > clientCtx.getConf().maxAllowedEnsembleChanges) {
             if (LOG.isDebugEnabled()) {
-                LOG.info("[EnsembleChange-L{}-{}] : resolved ledger metadata conflict and writing to zookeeper,"
-                        + " local meta data is \n {} \n, zk meta data is \n {}.",
-                        ledgerId, ensembleChangeIdx, metadata, newMeta);
+                LOG.debug("Ledger {} reaches max allowed ensemble change number {}",
+                          ledgerId, clientCtx.getConf().maxAllowedEnsembleChanges);
             }
-            // update znode version
-            metadata.setVersion(newMeta.getVersion());
-            // merge ensemble infos from new meta except last ensemble
-            // since they might be modified by recovery tool.
-            metadata.mergeEnsembles(newMeta.getEnsembles());
-            writeLedgerConfig(new ChangeEnsembleCb(ensembleInfo, curBlockAddCompletions,
-                    ensembleChangeIdx));
-            return true;
+            handleUnrecoverableErrorDuringAdd(WriteException);
+            return;
         }
 
-        @Override
-        public String toString() {
-            return String.format("ReReadLedgerMetadata(%d)", ledgerId);
-        }
+        new MetadataUpdateLoop(
+                clientCtx.getLedgerManager(), getId(),
+                this::getLedgerMetadata,
+                (metadata) -> {
+                    List<BookieSocketAddress> lastEnsemble = metadata.getLastEnsembleValue();
+                    boolean failedBookieInEnsemble = failedBookies.entrySet().stream()
+                        .anyMatch((e) -> lastEnsemble.get(e.getKey()).equals(e.getValue()));
+                    return !metadata.isClosed() && !metadata.isInRecovery() && failedBookieInEnsemble;
+                },
+                (metadata) -> {
+                    List<BookieSocketAddress> newEnsemble = EnsembleUtils.replaceBookiesInEnsemble(
+                            clientCtx.getBookieWatcher(), getId(), metadata, getCurrentEnsemble(), failedBookies);
+                    Long lastEnsembleKey = metadata.getLastEnsembleKey();
+                    LedgerMetadataBuilder builder = LedgerMetadataBuilder.from(metadata);
+                    long newEnsembleStartEntry = getLastAddConfirmed() + 1;
+                    checkState(lastEnsembleKey <= newEnsembleStartEntry,
+                               "New ensemble must either replace the last ensemble, or add a new one");
```

Comment:
Please print the lastEnsembleKey and newEnsembleStartEntry here; 
With this check state what happens? Does the write fail?
Also, when can this happen? If we are here we don't have any outstanding metadata update in the flight.
For the ledger L1 with Ensemble (B1, B2, B3) when LAC is at X;  X+1 write failed on Bookie B1. The ensemble is changed to (B4,B2,B3) and then we noticed that B2 also failed. In that case, do we end up in this situation? 
But that may be OK state right? why are you doing checkState?

Reply:
I believe that this cannot be violated if the protocol is operating correctly (the only other way for an ensemble to be a fence, but that's rules out by the previous argument).  The checkState call would appear to be a statement that this case is impossible?  The == case happens when we got a new failure since we sent the ensemble change (which worked), so we have to replace it.  @ivankelly Is that right?

Reply:
@jvrao @athanatos's understanding is correct. The checkState is an assertion.